File: src/api/quiz.ts
import apiClient from './client'

export enum QuizQuestionOrder {
    FIXED = 'FIXED',
    RANDOMIZED = 'RANDOMIZED'
}

export enum QuizResultVisibility {
    ALWAYS = 'ALWAYS',
    SCORE_ONLY = 'SCORE_ONLY',
    HIDDEN = 'HIDDEN'
}

export enum QuizEndType {
    MANUAL = 'MANUAL',
    SCHEDULED = 'SCHEDULED'
}

export interface QuizQuestionCreate {
    question_id: number
    position: number
    seed?: number | null
}

export interface QuizCreateRequest {
    name: string
    question_order: QuizQuestionOrder
    result_visibility: QuizResultVisibility
    end_type: QuizEndType
    questions: QuizQuestionCreate[]
}

export interface QuizResponse {
    id: string
    name: string
    teacher_id: string
    question_order: QuizQuestionOrder
    result_visibility: QuizResultVisibility
    end_type: QuizEndType
    created_at: string
    questions: Array<{
        id: string
        question_id: number
        position: number
        seed?: number | null
        question?: {
            id: number
            prompt: string
            type: string
            data?: Record<string, unknown>
            correct_answer?: Record<string, unknown>
            level?: number
        }
    }>
}

export interface QuizAssignmentCreate {
    quiz_id: string
    classroom_id?: string
    student_id?: string
    due_at?: string
    end_at?: string
}

export interface StudentQuizAssignmentResponse {
    id: string
    quiz_id: string
    quiz: QuizResponse
    due_at: string | null
    end_at: string | null
    created_at: string
}

export const quizApi = {
    createQuiz(payload: QuizCreateRequest) {
        return apiClient.post<{ data: QuizResponse }>('/teacher/quizzes', payload)
    },

    updateQuiz(quizId: string, payload: QuizCreateRequest) {
        return apiClient.put<{ data: QuizResponse }>(`/teacher/quizzes/${quizId}`, payload)
    },

    listQuizzes() {
        return apiClient.get<{ data: QuizResponse[] }>('/teacher/quizzes')
    },

    assignQuiz(payload: QuizAssignmentCreate) {
        return apiClient.post<{ data: unknown }>('/teacher/quizzes/assign', payload)
    },

    deleteQuiz(quizId: string) {
        return apiClient.delete<{ data: boolean }>(`/teacher/quizzes/${quizId}`)
    },

    listStudentAssignedQuizzes() {
        return apiClient.get<{ data: QuizResponse[] }>('/student/quizzes/all')
    }
}

File: backend/app/schemas/quiz.py
from __future__ import annotations

import uuid
from datetime import datetime
from pydantic import BaseModel, Field

from app.models.enums import QuizQuestionOrder, QuizResultVisibility, QuizEndType


from app.schemas.practice import QuestionPublic

class QuizQuestionBase(BaseModel):
    question_id: int
    position: int = 0


class QuizQuestionCreate(QuizQuestionBase):
    seed: int | None = None


class QuizQuestionResponse(QuizQuestionBase):
    id: uuid.UUID
    seed: int | None = None
    question: QuestionPublic | None = None

    class Config:
        from_attributes = True


class QuizBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    question_order: QuizQuestionOrder = QuizQuestionOrder.FIXED
    result_visibility: QuizResultVisibility = QuizResultVisibility.ALWAYS
    end_type: QuizEndType = QuizEndType.MANUAL


class QuizCreateRequest(QuizBase):
    questions: list[QuizQuestionCreate]


class QuizResponse(QuizBase):
    id: uuid.UUID
    teacher_id: uuid.UUID
    created_at: datetime
    questions: list[QuizQuestionResponse]

    class Config:
        from_attributes = True


class QuizAssignmentBase(BaseModel):
    quiz_id: uuid.UUID
    classroom_id: uuid.UUID | None = None
    student_id: uuid.UUID | None = None
    due_at: datetime | None = None
    end_at: datetime | None = None


class QuizAssignmentCreate(QuizAssignmentBase):
    pass


class QuizAssignmentResponse(QuizAssignmentBase):
    id: uuid.UUID
    created_at: datetime

    class Config:
        from_attributes = True


class StudentQuizAssignmentResponse(BaseModel):
    id: uuid.UUID
    quiz_id: uuid.UUID
    quiz: QuizResponse
    due_at: datetime | None = None
    end_at: datetime | None = None
    created_at: datetime

    class Config:
        from_attributes = True
